miio/gateway.py

```diff
@@ -529,45 +560,230 @@ def set_light(self, color_name, brightness):
         brightness_and_color = brightness_and_color_to_int(
             brightness, color_map[color_name]
         )
-        return self.send("set_rgb", [brightness_and_color])
+        return self._gateway.send("set_rgb", [brightness_and_color])
 
 
 class SubDevice:
-    def __init__(self, gw, sid, type_, _, __, ___):
-        self.gw = gw
+    def __init__(self, gw, sid, type, _, __, ___):
+        self._gw = gw
         self.sid = sid
-        self.type = DeviceType(type_)
+        self._battery = None
+        try:
+            self.type = DeviceType(type)
+        except ValueError:
+            self.type = DeviceType(-1)
 
-    def unpair(self):
-        return self.gw.send("remove_device", [self.sid])
+    def __repr__(self):
+        return "<Subdevice %s: %s fw: %s bat: %s>" % (
+            self.device_type,
+            self.sid,
+            self.get_firmware_version(),
+            self.get_battery(),
+        )
+
+    @property
+    def device_type(self):
+        return self.type.name
 
+    @property
     def battery(self):
-        return self.gw.send("get_battery", [self.sid])[0]
+        return self._battery
 
-    def get_firmware_version(self) -> Optional[int]:
-        """Returns firmware version"""
+    @command()
+    def subdevice_send(self, command):
+        """Send a command/query to the subdevice"""
         try:
-            return self.gw.send("get_device_prop", [self.sid, "fw_ver"])[0]
+            return self._gw.send(command, [self.sid])
         except Exception as ex:
-            _LOGGER.debug(
-                "Got an exception while fetching fw_ver: %s", ex, exc_info=True
+            _LOGGER.error(
+                "Got an exception while sending command %s: %s", command, ex, exc_info=True
             )
             return None
 
-    def __repr__(self):
-        return "<Subdevice %s: %s fw: %s bat: %s>" % (
-            self.type,
-            self.sid,
-            self.get_firmware_version(),
-            self.battery(),
-        )
+    @command()
+    def subdevice_send_arg(self, command, arguments):
+        """Send a command/query including arguments to the subdevice"""
+        try:
+            return self._gw.send(command, arguments, extra_parameters={"sid": self.sid})
+        except Exception as ex:
+            raise GatewayException("Got an exception while sending command '%s' with arguments '%s': %s" % (command, str(arguments), ex))
+
+    @command(click.argument("property"))
+    def get_subdevice_prop(self, property):
+        """Get the value of a property of the subdevice."""
+        try:
+            response = self._gw.send("get_device_prop", [self.sid, property])
+        except Exception as ex:
+            raise GatewayException("Got an exception while fetching property %s: %s" % (property, ex))
```

Comment:
`raise GatewayException(...) from ex`. I won't repeat this for for similar cases below, but please change them also to be consistent.

Reply:
done